### PR TITLE
Feature: Federation portals status

### PR DESCRIPTION
**Pull request by Bilelkihal**
## Changes
- Add a helper to store the status of federation portals in Rails cache to be used for
  - Federation portals input chips
  - Home page portals config
- The helper is initialized in Home, browse and search page in an asynch way to not affect pages waiting time. And for the federation input chips each chip loads separately within its own Turbo Frame, ensuring individual loading doesn't block other elements.
- The cache is valid for two hours

#### In the browse page
image

#### In search page
image

### CODE REVIEW

File: app/helpers/components_helper.rb [hunk 1]
```diff
@@ -292,4 +292,14 @@ def form_cancel_button
     end
   end
 
+  def chips_skelton
```
Comment (syphax-bouazzouni) [inline app/helpers/components_helper.rb:295]: this need to go the chip component haml file

File: app/views/home/federation_portals_status.html.haml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+= render TurboFrameComponent.new(id:"federation_portals_status_#{@key}") do |container|
+    %div{style: "cursor: default;"}
+        = group_chip_component(name: @name, object: { "acronym" => @acronym, "value" => @key }, checked: @checked, title: "#{!@portal_up ? "#{@key.humanize.gsub('portal', 'Portal')} is currently down" : ''}", disabled: !@portal_up)
+
+-# TODO put the portal is down tooltip message in local lang files
```
Comment (syphax-bouazzouni) [inline app/views/home/federation_portals_status.html.haml:5]: todo the todo

File: app/views/ontologies/browser/browse.html.haml [hunk 3]
```diff
@@ -73,9 +73,11 @@
                                 = render LoaderComponent.new(small:true)
           .browse-federation-input-chip-container                    
             = render DropdownContainerComponent.new(id: "browse-portal-filter", is_open: !request_portals.empty?, title: t('federation.results_from_external_portals')) do
-              .browse-filter-checks-container.px-1
+              .px-1.browse-federation-input-chips
                 - federated_portals.each do |key, config|
-                  = group_chip_component(name: "portals", object: { "acronym" => config[:name], "value" => key }, checked: request_portals.include?(key.to_s), title: '')
+                  = render TurboFrameComponent.new(id: "federation_portals_status_#{key}", src: "home/federation_portals_status?name=portals&key=#{key}&acronym=#{config[:name]}&checked=#{request_portals.include?(key.to_s)}") do |container|
```
Comment (syphax-bouazzouni) [inline app/views/ontologies/browser/browse.html.haml:78]: this need to be a helper as reused in search index page

File: config/routes.rb [hunk 4]
```diff
@@ -131,6 +131,7 @@
   end
 
   get '' => 'home#index'
+  get 'home/federation_portals_status', to: 'home#federation_portals_status'
```
Comment (syphax-bouazzouni) [inline config/routes.rb:134]: this route  needs to be renamed to something  like this: `/status/:portal_name`

File: app/views/search/index.html.haml [hunk 5]
```diff
@@ -20,7 +20,9 @@
                 = t('federation.results_from_external_portals')
               .field.d-flex 
                 - federated_portals.each do |key, config|
-                  = group_chip_component(name: "portals[]", object: { "acronym" => config[:name], "value" => key }, checked: request_portals.include?(key.to_s), title: '')
+                  = render TurboFrameComponent.new(id: "federation_portals_status_#{key}", src: "home/federation_portals_status?name=portals[]&key=#{key}&acronym=#{config[:name]}&checked=#{request_portals.include?(key.to_s)}") do |container|
```
Comment (syphax-bouazzouni) [inline app/views/search/index.html.haml:23]: this need to be a helper as reused in browse  page

File: app/views/home/index.html.haml [hunk 6]
```diff
@@ -230,6 +230,13 @@
         %a{href:logo[:url], target: "_blanc"}
           %img{src: asset_path(logo[:img_src])}
 
+-# This section is hidden, used to initialize the cache of federation portals status
+.d-none
+  - federated_portals.each do |key, config|
+    = render TurboFrameComponent.new(id: "federation_portals_status_#{key}", src: "home/federation_portals_status?name=portals[]&key=#{key}&acronym=#{config[:name]}&checked=#{request_portals.include?(key.to_s)}") do |container|
```
Comment (syphax-bouazzouni) [inline app/views/home/index.html.haml:236]: this need to be a helper as reused in the search and browse  page

File: app/views/home/index.html.haml [hunk 7]
```diff
@@ -230,6 +230,13 @@
         %a{href:logo[:url], target: "_blanc"}
           %img{src: asset_path(logo[:img_src])}
 
+-# This section is hidden, used to initialize the cache of federation portals status
```
Comment (syphax-bouazzouni) [inline app/views/home/index.html.haml:233]: always replace comments with a better function names (a good video about it https://www.youtube.com/watch?v=wXaC0YvDgIo&list=PL9wALaIpe0Py6E_oHCgTrD6FvFETwJLlx)

File: app/helpers/home_helper.rb [hunk 8]
```diff
@@ -23,12 +23,12 @@ def format_number_abbreviated(number)
   end
 
   def portal_config_tooltip(portal_name, &block)
-    title = render TurboFrameComponent.new(id: "portal_config_tooltip_#{portal_name&.downcase}", src: "/config?portal=#{portal_name&.downcase}", style: "width: 600px !important; max-height: 300px; overflow: scroll")
+    title = federation_portal_status(portal_name: portal_name.downcase) ? render(TurboFrameComponent.new(id: "portal_config_tooltip_#{portal_name&.downcase}", src: "/config?portal=#{portal_name&.downcase}", style: "width: 600px !important; max-height: 300px; overflow: scroll")) : nil
```
Comment (syphax-bouazzouni) [inline app/helpers/home_helper.rb:26]: for readability not using a ternary would be  better

```
def portal_config_tooltip(portal_name, &block)
  portal_id = portal_name&.downcase

  title = if federation_portal_status(portal_name: portal_id)
    render(
      TurboFrameComponent.new(
        id: "portal_config_tooltip_#{portal_id}",
        src: "/config?portal=#{portal_id}",
        style: "width: 600px !important; max-height: 300px; overflow: scroll"
      )
    )
  end

  render Display::InfoTooltipComponent.new(text: title, interactive: true) do
    capture(&block)
  end
end

```

File: app/views/home/federation_portals_status.html.haml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+= render TurboFrameComponent.new(id:"federation_portals_status_#{@key}") do |container|
```
Comment (syphax-bouazzouni) [inline app/views/home/federation_portals_status.html.haml:1]: no need for this file you can render it directly in the controller, 

soothing like this 

```ruby 
render inline: helpers.your_helper
```

one less file is always a win, future us.

**Review by syphax-bouazzouni — CHANGES_REQUESTED** [review]

File: app/views/home/index.html.haml (hunk 7 above)
Comment (Bilelkihal) [reply app/views/home/index.html.haml]: Thank you

**Review by Bilelkihal — COMMENTED** [review]
✅

File: app/helpers/federation_helper.rb [hunk 10]
```diff
@@ -124,4 +124,59 @@ def federation_external_class?(class_object)
...
+        end.head
+        portal_up = response.success?
+      rescue StandardError => e
+        Rails.logger.error("Error checking portal status for #{portal_name}: #{e.message}")
+      end
+      portal_up
+    end
+  end
+
+  def federation_portals_status(key,name,acronym,checked,portal_up)
```
Comment (Bilelkihal) [inline app/helpers/federation_helper.rb:147]: This function and the one in the line 127 have a very close name, we should change the name of this to something like:
`federation_chip_component`

Comment (syphax-bouazzouni) [reply app/helpers/federation_helper.rb:147]: ok

Comment (syphax-bouazzouni) [reply app/helpers/federation_helper.rb]: done in https://github.com/ontoportal-lirmm/bioportal_web_ui/pull/769/commits/321b81747cfebeba66dee98143227ddc4b4d9fae

File: app/helpers/components_helper.rb (hunk 1 above)
Comment (syphax-bouazzouni) [reply app/helpers/components_helper.rb:295]: done in https://github.com/ontoportal-lirmm/bioportal_web_ui/pull/769/commits/608c867bb548ac97e70c2d9313338faaf7188889

**Review by syphax-bouazzouni — APPROVED** [review]

**Pull request merged**